src/v/archival/service.h

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+#include "archival/manifest.h"
+#include "archival/ntp_archiver_service.h"
+#include "cluster/partition_manager.h"
+#include "model/fundamental.h"
+#include "s3/client.h"
+#include "storage/api.h"
+#include "storage/log_manager.h"
+#include "storage/ntp_config.h"
+#include "storage/segment.h"
+#include "storage/segment_set.h"
+#include "utils/intrusive_list_helpers.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/weak_ptr.hh>
+
+#include <absl/container/node_hash_map.h>
+
+#include <queue>
+
+namespace archival {
+namespace internal {
+
+class ntp_upload_queue {
+    struct upload_queue_item {
+        ss::lw_shared_ptr<ntp_archiver> archiver;
+        intrusive_list_hook _upl_hook;
```

Comment:
do you need to disable move operations on this type, or implement move so that it calls `intrusive_list_hook::swap` on the nodes? IIRC this is needed to make them move safe, but it's been a while since i thought about it.